### PR TITLE
builtins: initialize FD halo

### LINKED ISSUE #1177: Reorganize builtins
- create builtins subdirectory
- sort functions into `builtins/maths.py` , `builtins/misc.py`
- potentially, split `test_builtins.py` accordingly: `test_builtins_maths.py` `test_builtins_misc.py`
- add rand for `SparseFunctions`, similar to `norm`, `mmin`, `mmax`

### LINKED ISSUE #1250: Non-padding of the halo could cause incorrect computed values
The following example results in NaN values.

```python

    grid = Grid(shape=(5,5))
 
    v = VectorTimeFunction(name="v", grid=grid, time_order=1, space_order=4)
    p = TimeFunction(name="p2", grid=grid, time_order=1, space_order=4, staggered=NODE)

    v[0].data[:] = 5.
    v[1].data[:] = 5.
 
    den = np.zeros((5,5), dtype=np.float32)
    den[:] = 2.

    B = Function(name="B", grid=grid, space_order=4, staggered=NODE)
    initialize_function(B, den, 0, dtype=np.float32)
    b = 1./B

    eq = Eq(p, div(b*v))
    op = Operator([eq])
    op.apply(time_M=0)
```

 
This NaN values are coming from non-padding of the halo in B. More precisely this occurs because the halo, non padded, is used in the derivative on B leading to 1/0.

As suggest by @mloubout padding the B as ` B._data_with_outhalo[:] = 2 ` solve this problem manually for this case.

### PR DESCRIPTION
**Pull request by mloubout**
- Fixes #1250 
- Fixes #1177
- Adds shifted div/grad

### CODE REVIEW

File: devito/builtins.py [hunk 1]
```diff
@@ -255,6 +255,8 @@ def initialize_function(function, data, nbl, mapper=None, mode='constant',
         accepted.
     name : str, optional
         The name assigned to the operator.
+    pad_halo : bool, optional
+        Whether to pad the FD halo as well.
```
Comment (rhodrin) [inline devito/builtins.py:259]: *Whether to also pad the ...

File: devito/builtins.py [hunk 2]
```diff
@@ -364,6 +367,29 @@ def buff(i, j):
 
     assign(lhs, rhs, options=options, name=name, **kwargs)
 
+    if pad_halo:
+        pad_fd_halo(function)
+
+
+def pad_fd_halo(function):
```
Comment (rhodrin) [inline devito/builtins.py:374]: Is `builtins.py` the right place for this?

File: tests/test_builtins.py [hunk 3]
```diff
@@ -239,6 +239,37 @@ def test_if_parallel(self):
             assert np.all(a[::-1, 3:6] - np.array(f.data[12:18, 9:12]) == 0)
             assert np.all(a[3:6, ::-1] - np.array(f.data[9:12, 12:18]) == 0)
 
+    @pytest.mark.parametrize('ndim, nbl', [
```
Comment (rhodrin) [inline tests/test_builtins.py:242]: I think we should have a test with mpi also?

File: devito/builtins.py [hunk 4]
```diff
@@ -364,6 +367,29 @@ def buff(i, j):
...
+    """ Pad finite-difference halo with edge values."""
+    h_shape = function._data_with_outhalo.shape
+    for i, h in enumerate(function._size_outhalo):
+        if h.left == 0 and h.right == 0:
+            continue
+        slices = [slice(None)]*len(function.shape)
+        slices_d = [slice(None)]*len(function.shape)
+        # left part
+        slices[i] = slice(0, h.left)
```
Comment (rhodrin) [inline devito/builtins.py:383]: this works ok with mpi where on a rank you could end up with
```
        slices[i] = slice(0, 0)
        slices_d[i] = slice(0, 1, 1)
```
?

Comment (rhodrin) [reply devito/builtins.py:383]: Same for the right bit.

File: devito/builtins.py (hunk 2 above)
Comment (mloubout) [reply devito/builtins.py:374]: It's just a local function because it's used at two different places, similar to the `nbl` setup function. Not sure where else would put it.

File: tests/test_builtins.py (hunk 3 above)
Comment (mloubout) [reply tests/test_builtins.py:242]: Yes we should, working on it

File: devito/builtins.py (hunk 4 above)
Comment (mloubout) [reply devito/builtins.py:383]: Not sure I follow

File: devito/builtins.py (hunk 1 above)
Comment (mloubout) [reply devito/builtins.py]: Fixed

File: tests/test_builtins.py (hunk 3 above)
Comment (mloubout) [reply tests/test_builtins.py:242]: Added MPI test

File: devito/builtins.py (hunk 4 above)
Comment (mloubout) [reply devito/builtins.py]: Fixed MPI case

File: .github/workflows/tutorials.yml [hunk 5]
```diff
@@ -67,7 +67,7 @@ jobs:
       if: runner.os == 'macOS'
       run: |
         if [ "${{ matrix.compiler }}" = "gcc-9" ]; then
-          brew install gcc
+          brew install gcc@9
```
Comment (FabioLuporini) [inline .github/workflows/tutorials.yml:70]: out of curiosity, why the change?

File: devito/builtins.py [hunk 6]
```diff
@@ -255,6 +255,8 @@ def initialize_function(function, data, nbl, mapper=None, mode='constant',
         accepted.
     name : str, optional
         The name assigned to the operator.
+    pad_halo : bool, optional
+        Whether to also pad the finite-difference halo.
```
Comment (FabioLuporini) [inline devito/builtins.py:259]: Drop "finite-difference" or replace with something else

File: devito/builtins.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/builtins.py:374]: yeah, we need to move the builtins to /builtins such that we can also have a utils.py

since it's a private function, the name could start with a _

File: devito/builtins.py [hunk 7]
```diff
@@ -364,6 +367,29 @@ def buff(i, j):
 
     assign(lhs, rhs, options=options, name=name, **kwargs)
 
+    if pad_halo:
+        pad_fd_halo(function)
+
+
+def pad_fd_halo(function):
+    """ Pad finite-difference halo with edge values."""
```
Comment (FabioLuporini) [inline devito/builtins.py:375]: drop "finite-difference"

File: devito/finite_differences/differentiable.py [hunk 8]
```diff
@@ -240,16 +240,17 @@ def laplace(self):
         derivs = tuple('d%s2' % d.name for d in space_dims)
         return Add(*[getattr(self, d) for d in derivs])
 
-    @property
-    def div(self):
+    def div(self, shift=None):
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py:243]: turning this property into a function -- should we consider it a major API change?

File: devito/finite_differences/differentiable.py [hunk 9]
```diff
@@ -240,16 +240,17 @@ def laplace(self):
...
         space_dims = [d for d in self.dimensions if d.is_Space]
-        derivs = tuple('d%s' % d.name for d in space_dims)
-        return Add(*[getattr(self, d) for d in derivs])
+        return Add(*[getattr(self, 'd%s' % d.name)(x0=None if shift is None else
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py:245]: easier to read as

```
x0=shift or (d + shift * d.spacing)
```

?

File: devito/finite_differences/differentiable.py [hunk 10]
```diff
@@ -240,16 +240,17 @@ def laplace(self):
...
+                                                   d + shift * d.spacing)
+                     for d in space_dims])
 
-    @property
-    def grad(self):
+    def grad(self, shift=None):
         from devito.types.tensor import VectorFunction, VectorTimeFunction
-        comps = [getattr(self, 'd%s' % d.name) for d in self.dimensions if d.is_Space]
+        comps = [getattr(self, 'd%s' % d.name)(x0=None if shift is None else
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py:251]: same as above?

File: devito/builtins.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/builtins.py:374]: > yeah, we need to move the builtins to /builtins such that we can also have a utils.py

this also appears as an issue on GH. We could do it now

Comment (mloubout) [reply devito/builtins.py:374]: Doable

File: .github/workflows/tutorials.yml (hunk 5 above)
Comment (mloubout) [reply .github/workflows/tutorials.yml:70]: brew now installs gcc 10 by default, was easier to fix just this line than the other ones for gcc 10.

File: devito/finite_differences/differentiable.py (hunk 8 above)
Comment (mloubout) [reply devito/finite_differences/differentiable.py:243]: No this is not user facing, `div(f)` is and hasn't changed.

File: devito/finite_differences/differentiable.py (hunk 9 above)
Comment (mloubout) [reply devito/finite_differences/differentiable.py:245]: That's also not the same and wrong.

`None if shift is None else  (d + shift * d.spacing)` 
- None => None
- value => d + value * spacing

`x0=shift or (d + shift * d.spacing)` : 
- None => d + None * d.spacing , that's not valid
- value => value

File: devito/finite_differences/differentiable.py (hunk 10 above)
Comment (mloubout) [reply devito/finite_differences/differentiable.py:251]: same

File: devito/builtins.py (hunk 6 above)
Comment (mloubout) [reply devito/builtins.py:259]: But that's the FD halo from the space order, why droping it?

File: devito/builtins.py (hunk 2 above)
Comment (mloubout) [reply devito/builtins.py]: Done

File: devito/finite_differences/differentiable.py (hunk 9 above)
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py:245]: ah, LOL, don't know what my mind was processing when I wrote that comment above, ignore me :grimacing:

**Review by FabioLuporini — COMMENTED** [review]
I really like the new `builtins/`

File: devito/builtins.py (hunk 6 above)
Comment (FabioLuporini) [reply devito/builtins.py:231]: the halo is a concept that applies to any (un)structured grid computation, so it's broader than FD

File: tests/test_builtins.py [hunk 11]
```diff
@@ -239,6 +239,72 @@ def test_if_parallel(self):
...
+            assert np.take(f._data_with_outhalo, -1, axis=1)[7, 7] == 6
+        else:
+            assert f._data_with_outhalo[7, 7] == 0
+            assert np.take(f._data_with_outhalo, 0, axis=-1)[7] == 1
+            assert np.take(f._data_with_outhalo, -1, axis=-1)[7] == 3
+
+    @skipif('nompi')
+    @pytest.mark.parametrize('nbl', [0, 2])
+    @pytest.mark.parallel(mode=4)
+    def test_if_halo_mpi(self, nbl):
```
Comment (FabioLuporini) [inline tests/test_builtins.py:278]: I don't know why, but it looks like my comment about merging this and the non-MPI test into one test with `pytest.mark.parallel(mode=[1, 2, 4])` has disappeared?

Comment (mloubout) [reply tests/test_builtins.py:278]: `mode=1` is still MPI with one rank, not non-mpi.

File: devito/builtins.py (hunk 6 above)
Comment (mloubout) [reply devito/builtins.py]: Changed to outer halo

**Review by FabioLuporini — APPROVED** [review]
GTG, thanks

**Review by rhodrin — APPROVED** [review]
